mantatail.py

```diff
@@ -69,33 +69,44 @@ def handle_motd(self, user_nick):
         start_msg = bytes(motd_start_and_end["start_msg"], encoding=self.encoding)
         end_msg = bytes(motd_start_and_end["end_msg"], encoding=self.encoding)
 
-        self.client_socket.sendall(start_msg)
+        self.user.socket.sendall(start_msg)
 
         for motd_line in motd:
             motd_msg = bytes(
-                f"{self.send_to_client_prefix} {motd_num} {user_nick} :{motd_line}{self.send_to_client_suffix}",
+                f"{self.send_to_client_prefix} {motd_num} {self.user.nick} :{motd_line}{self.send_to_client_suffix}",
                 encoding=self.encoding,
             )
-            self.client_socket.sendall(motd_msg)
+            self.user.socket.sendall(motd_msg)
 
-        self.client_socket.sendall(end_msg)
+        self.user.socket.sendall(end_msg)
 
     def handle_join(self, message):
-        channel_regex = r"[&#+!][^ \x07,]{1,49}"  # Covers max 200 characters?
+        channel_regex = r"#[^ \x07,]{1,49}"  # TODO: Make more restrictive (currently valid: ###, #ö?!~ etc)
+
         if not re.match(channel_regex, message):
-            self.client_socket.sendall(
-                bytes(f"{self.send_to_client_prefix} ", encoding=self.encoding)
+            no_channel_num, no_channel_info = (
+                irc_response_nums["error_replies"]["ERR_NOSUCHCHANNEL"][0],
+                irc_response_nums["error_replies"]["ERR_NOSUCHCHANNEL"][1].replace(
+                    "<channel name>", message
+                ),
             )
+
+            self.generate_error_reply(no_channel_num, no_channel_info)
+
         else:
-            server.channels[message] = Channel()
-            # print("CHANNELS", server.channels)
+            if not message in self.server.channels.keys():
+                self.server.channels[message] = Channel()
+
+            if self.user.nick in self.server.channels[message].user_list.keys():
+                return
+            else:
+                self.server.channels[message].user_list[self.user.nick] = self.user
```

Comment:
This is currently quite repetitive. You can use `not in`:

```suggestion
            if self.user.nick not in self.server.channels[message].user_list.keys():
                self.server.channels[message].user_list[self.user.nick] = self.user
```

Or just unconditionally add the user, it doesn't matter if it's already added:

```suggestion
            self.server.channels[message].user_list[self.user.nick] = self.user
```